### DISCUSSION TITLE
middleware breaks behavior of background task

### DISCUSSION THREAD

**Discussion by jkh911208** [fastapi/fastapi#7168] [Questions] [answered]
### First check

* [x ] I added a very descriptive title to this issue.
* [ x] I used the GitHub search to find a similar issue and didn't find it.
* [ x] I searched the FastAPI documentation, with the integrated search.
* [ x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [ x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [ x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [ x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

```Python
import asyncio
import logging
import time

from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
from starlette.background import BackgroundTasks

app = FastAPI()


@app.get("/")
async def read_root():
    response_object = {"Hello": "World"}
    tasks = BackgroundTasks()
    tasks.add_task(bg_task)
    return JSONResponse(response_object, 202, background=tasks)


async def bg_task():
    logging.error("test1")
    await asyncio.sleep(10)
    logging.error("test2")


@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
    start_time = time.time()
    response = await call_next(request)
    response.headers["X-Process-Time"] = str(time.time() - start_time)
    response.headers["address"] = request.client.host
    return response

if __name__ == "__main__":
    import uvicorn
    uvicorn.run("main:app", host="0.0.0.0", port=8000,
                log_level="debug", reload=True)

```

### Description

run the above code, make a request to "/"
you will see 
<img width="520" alt="Screen Shot 2020-10-21 at 8 50 42 AM" src="https://user-images.githubusercontent.com/17028674/96729219-89934500-137a-11eb-93a6-70e0bdd49ece.png">
which logging is happening before return the object

but if you make only one request at a time this is fine, but if you make multiple request while background task is running

problem #1 the second request start after first background task finishes
problem #2 when background task of second request finishes, it throw error
<img width="993" alt="Screen Shot 2020-10-21 at 8 52 39 AM" src="https://user-images.githubusercontent.com/17028674/96729600-f1499000-137a-11eb-8cdb-5e4d66c11c5c.png">

### Environment

* OS: [e.g. Linux / Windows / macOS]: macos and debian
* FastAPI Version [e.g. 0.3.0]: 0.61.1
* Python version: 3.8.5

### Additional context

if i remove middleware code it works as expected
<img width="592" alt="Screen Shot 2020-10-21 at 8 55 24 AM" src="https://user-images.githubusercontent.com/17028674/96729820-2d7cf080-137b-11eb-961f-fc655a448073.png">
all requests are handled and all background tasks are working fine no exception

**Comment by ycd** [comment]
I believe this is related to Starlette Issue 919:  [Background tasks don't work with middleware that subclasses `BaseHTTPMiddleware`](https://github.com/encode/starlette/issues/919)

**Comment by jkh911208** [comment]
thank you @ycd seems like work is still in progress

**Comment by Kludex** [comment]
@jkh911208 Could you close this issue, please?

**Comment by yingyingPengSAP** [comment]
For those who are searching for how to call background tasks under FastAPI middleware, here is quick solution: 

from starlette.background import BackgroundTask
from somewhere import functionA

@app.middleware("http")
async def middleware(request: Request, call_next):
        ........(do something)
        response.background = BackgroundTask(functionA, arg)
    return response

You can add a time.sleep(10) in function A to see the result better. 
For tasks running in background, it will return the results of function A
very quickly and will not be affected by time.sleep().

Hope this helps someone!

**Comment by santigandolfo** [comment]
Hi @jkh911208  @Kludex , why is this issue `Closed`? Cause I'm trying to use background tasks and a middleware at the same time, and if I send a request to the endpoint through Postman, I receive my response but the request time doesn't finish until the background task does. But if I remove the middleware then it works as expected.

**Comment by Kludex** [comment]
What is the version of your FastAPI?

Oh sorry, yeah, this is still a thing. https://github.com/encode/starlette/pull/1441

**Comment by santigandolfo** [comment]
> What is the version of your FastAPI?
> 
> Oh sorry, yeah, this is still a thing. [encode/starlette#1441](https://github.com/encode/starlette/pull/1441)

I'm on v0.75.2
Do you suggest a workaround until that PR is merged in Starlette and FastAPI bumps it's requirements?

**Comment by Kludex** [comment]
Yes, don't use `BaseHTTPMiddleware`.

**Comment by santigandolfo** [comment]
> Yes, don't use `BaseHTTPMiddleware`.

In that case I couldn't add general before and after request actions, right?

**Comment by Kludex** [comment]
No, I'm saying to not use that class. You can have middlewares: https://pgjones.dev/blog/how-to-write-asgi-middleware-2021

**Comment by Kludex** [comment, marked as answer]
This is solved on the latest FastAPI.

**Reply by Molkree** [reply]
@Kludex, I still experience this issue on the latest FastAPI ([0.95.0](https://github.com/tiangolo/fastapi/releases/tag/0.95.0)) (without the error in the end, just sequential behavior).

The last "good" version seems to be [0.90.0](https://github.com/tiangolo/fastapi/releases/tag/0.90.0), the first "bad" version is [0.90.1](https://github.com/tiangolo/fastapi/releases/tag/0.90.1) (with a caveat). That version [allowed](https://github.com/tiangolo/fastapi/pull/5980) starlette [0.23.1](https://github.com/encode/starlette/releases/tag/0.23.1). The only change [between](https://github.com/encode/starlette/compare/0.23.0...0.23.1) starlette 0.23.0 and 0.23.1 seems to be [this PR](https://github.com/encode/starlette/pull/1940).

The caveat about FastAPI 0.90.1 is that it allows starlette 0.23.0 and 0.23.1 so FastAPI 0.90.1/starlette 0.23.0 still works correctly but FastAPI 0.90.1/starlette 0.23.1 doesn't. FastAPI>=0.91.0 requires starlette>=0.24.0.

Just to clarify, the original code sample from 2020 doesn't work correctly with starlette>=0.23.1. But here's my slightly different example with time logging that I run with `uvicorn test_fastapi:app --reload`:
```py
import logging
from time import sleep
from typing import Any

from fastapi import BackgroundTasks, FastAPI, Request
from fastapi import status as status_code

logging.basicConfig(
    level=logging.INFO,
    format="[%(asctime)s] [%(levelname)s] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
)

app = FastAPI()


@app.middleware("http")
async def http_middleware(request: Request, call_next: Any) -> Any:
    return await call_next(request)


@app.post("/task", status_code=status_code.HTTP_201_CREATED, summary="Create task")
def create_task(background_tasks: BackgroundTasks) -> None:
    logging.info("Entered POST method")
    background_tasks.add_task(background_task)


def background_task() -> None:
    logging.info("Started background task")
    sleep(10)
    logging.info("Finished background task")

```

Here's how it looks with FastAPI 0.90.1/starlette 0.23.0:
```
INFO:     Application startup complete.
[2023-03-22 04:37:22] [INFO] Entered POST method
INFO:     127.0.0.1:52762 - "POST /task HTTP/1.1" 201 Created
[2023-03-22 04:37:22] [INFO] Started background task
[2023-03-22 04:37:23] [INFO] Entered POST method
INFO:     127.0.0.1:52762 - "POST /task HTTP/1.1" 201 Created
[2023-03-22 04:37:23] [INFO] Started background task
[2023-03-22 04:37:33] [INFO] Finished background task
[2023-03-22 04:37:33] [INFO] Finished background task
```

Here's how it looks with FastAPI 0.90.1/starlette 0.23.1 (or FastAPI 0.95.0/starlette 0.26.1):
```
INFO:     Application startup complete.
[2023-03-22 04:39:22] [INFO] Entered POST method
INFO:     127.0.0.1:53173 - "POST /task HTTP/1.1" 201 Created
[2023-03-22 04:39:22] [INFO] Started background task
[2023-03-22 04:39:32] [INFO] Finished background task
[2023-03-22 04:39:32] [INFO] Entered POST method
INFO:     127.0.0.1:53173 - "POST /task HTTP/1.1" 201 Created
[2023-03-22 04:39:32] [INFO] Started background task
[2023-03-22 04:39:42] [INFO] Finished background task
```

**Reply by Kludex** [reply]
I can't reproduce the behavior you mentioned.

**Reply by Molkree** [reply]
@Kludex, that's unfortunate. What other info would help to pinpoint the issue? I have tested this on Windows with Python 3.8 but it was initially encountered on Ubuntu with Python 3.8 so both OSes should be affected. (Maybe macOS is fine?) Not sure if `uvicorn` version could affect this, I think I had latest version but I'll double check.

Edit: I've clarified above that I don't see the thrown error like in the original report, just sequential execution.

**Reply by Kludex** [reply]
Are you using the latest uvicorn? Can you try on a different Python version?

**Reply by Molkree** [reply]
@Kludex, yes, latest uvicorn ([0.21.1](https://github.com/encode/uvicorn/releases/tag/0.21.1)).
I've tried Python 3.8/11 on Windows and Python 3.7/8/10/11 on Ubuntu, always getting the same sequential execution :(

**Reply by santigandolfo** [reply]
@Kludex I just copied what @Molkree did running and python 3.11.2 and I'm being able to replicate the problem:
```
[2023-03-22 13:21:05] [INFO] Entered POST method
INFO:     127.0.0.1:52058 - "POST /task HTTP/1.1" 201 Created
[2023-03-22 13:21:05] [INFO] Started background task
[2023-03-22 13:21:15] [INFO] Finished background task
[2023-03-22 13:21:15] [INFO] Entered POST method
INFO:     127.0.0.1:52058 - "POST /task HTTP/1.1" 201 Created
[2023-03-22 13:21:15] [INFO] Started background task
[2023-03-22 13:21:25] [INFO] Finished background task
```

This is my `pipenv graph`:
```
fastapi==0.95.0
  - pydantic [required: >=1.6.2,<2.0.0,!=1.8.1,!=1.8,!=1.7.3,!=1.7.2,!=1.7.1,!=1.7, installed: 1.10.6]
    - typing-extensions [required: >=4.2.0, installed: 4.5.0]
  - starlette [required: >=0.26.1,<0.27.0, installed: 0.26.1]
    - anyio [required: >=3.4.0,<5, installed: 3.6.2]
      - idna [required: >=2.8, installed: 3.4]
      - sniffio [required: >=1.1, installed: 1.3.0]
```
But if I also set my requirements to have
```
fastapi==0.90.1
starlette==0.23.0
```

I get the same result.

The only way I can make it "work" is by commenting out the middleware (which is not what we want)

**Reply by Kludex** [reply]
Is it reproducible only with Starlette? If yes, can you create an issue on Starlette? I'll investigate.

**Reply by Molkree** [reply]
@Kludex, I can't reproduce it only with Starlette.

Here's my attempt:
test_starlette.py
```py
import logging
from time import sleep

from starlette.applications import Starlette
from starlette.background import BackgroundTasks
from starlette.middleware import Middleware
from starlette.middleware.base import BaseHTTPMiddleware, RequestResponseEndpoint
from starlette.requests import Request
from starlette.responses import JSONResponse, Response
from starlette.routing import Route

logging.basicConfig(
    level=logging.INFO,
    format="[%(asctime)s] [%(levelname)s] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
)


class CustomMiddleware(BaseHTTPMiddleware):
    async def dispatch(
        self, request: Request, call_next: RequestResponseEndpoint
    ) -> Response:
        return await call_next(request)


def create_task(request: Request) -> JSONResponse:
    logging.info("Entered POST method")
    tasks = BackgroundTasks()
    tasks.add_task(background_task)
    message = {"status": "Task created"}
    return JSONResponse(message, background=tasks, status_code=201)


def background_task() -> None:
    logging.info("Started background task")
    sleep(10)
    logging.info("Finished background task")


routes = [Route("/task", endpoint=create_task, methods=["POST"])]
middleware = [Middleware(CustomMiddleware)]
app = Starlette(routes=routes, middleware=middleware)
```
starlette-requirements.txt
```
starlette==0.26.1
uvicorn[standard]==0.21.1
```
Ran with `uvicorn test_starlette:app --reload`.
Just sent curl requests one after another with `curl -X 'POST' 'http://127.0.0.1:8000/task'`.
```
INFO:     Application startup complete.
[2023-03-23 02:20:27] [INFO] Entered POST method
INFO:     127.0.0.1:53582 - "POST /task HTTP/1.1" 201 Created
[2023-03-23 02:20:27] [INFO] Started background task
[2023-03-23 02:20:27] [INFO] Entered POST method
INFO:     127.0.0.1:53583 - "POST /task HTTP/1.1" 201 Created
[2023-03-23 02:20:27] [INFO] Started background task
[2023-03-23 02:20:37] [INFO] Finished background task
[2023-03-23 02:20:37] [INFO] Finished background task
```

Working as expected.

**Reply by Molkree** [reply]
@Kludex, alright, I have a clue!

After rewriting the example for Starlette I just used curl, not Swagger UI. The thing is, using curl with FastAPI not causes any issues too!

Just a TLDR: take code from my first message + this `requirements.txt`:
```
fastapi
starlette==0.23.1  # or >0.23.0
uvicorn[standard]==0.21.1
```
Run with `uvicorn test_fastapi:app --reload`. Open Swagger UI (http://127.0.0.1:8000/docs), and try executing the only POST method twice in succession. The second time will stall for ~10 seconds. Sending the same POST request with other means (curl) works as expected with no stalling.

If you pin Starlette to 0.23.**0** instead, executing from Swagger UI won't stall.

Could you confirm you are seeing the same behavior, @Kludex @santigandolfo?

@Kludex, do you think it's a bug with Swagger UI? Or Starlette [no longer](https://github.com/encode/starlette/pull/1940) behaving like it should in this case?

**Reply by Kludex** [reply]
Ah, the issue happens with the same connection. That's why you see it on Swagger UI - it uses the same connection for both requests. You can also reproduce with:

```python
from httpx import Client

with Client(base_url="http://localhost:8000") as client:
    client.post("/task")
    client.post("/task")
```

I'm not sure how no one noticed this in 4 months. :disappointed: 

Would you mind creating an issue on Starlette?

**Reply by Molkree** [reply]
@Kludex, opened https://github.com/encode/starlette/issues/2093

**Reply by Kludex** [reply]
Thanks.
